Design-Docs/standardizing-cluster-infrastructure.md

```diff
@@ -0,0 +1,277 @@
+
+To achieve this, apart from defining cluster stacks as described in Proposal 1, the provider integrations of Cluster API have to be aligned. SCS should define certain parameters, e.g. `ProviderMachineTemplate`, and all provider integrations of SCS providers should have some kind of translation matrix so that this input can be converted into the relevant provider object, e.g. `HCloudMachineTemplate`.
+
+We want to leave out the details of how this translation from SCS-defined properties into the respective provider integrations works. It can be discussed in future meetings.
+
+### Proposal 3: Applications are configured independently of the cluster infrastructure
+
+The first step of achieving independence of the layers, the cluster stacks have to be aligned in a certain way, e.g. that a CSR controller is a must-have for running metrics-server securely. This has been described in Proposal 1.
+
+The second step consists of two tasks:
+
+- Defining an adapted SCS cloud provider interface that unifies the inputs for creating load balancers
+- Standardizing the naming of storage classes
+
+A normal CCM has one object that it creates: the load balancer. Based on a service object that is defined by the user and that contains certain annotations, the CCM creates a load balancer. Service objects are written by users and play a big role in a lot of applications. Right now, if a user wants to switch from one cloud provider to another, he needs to adapt the annotations he sets in the service objects of his application configurations.
+
+Example:
+
+- OpenStack load balancer: [https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/openstack/loadbalancer.go#L67](https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/openstack/loadbalancer.go#L67)
+- Hetzner load balancer: [https://github.com/syself/hetzner-cloud-controller-manager/blob/master/internal/annotation/load_balancer.go](https://github.com/syself/hetzner-cloud-controller-manager/blob/master/internal/annotation/load_balancer.go)
+
+This can be avoided rather easily. We have to implement a new interface for load balancers, as a counterpart to the current one ([https://pkg.go.dev/k8s.io/cloud-provider#LoadBalancer](https://pkg.go.dev/k8s.io/cloud-provider#LoadBalancer)), which is a part of the cloud provider interface ([https://pkg.go.dev/k8s.io/cloud-provider#Interface](https://pkg.go.dev/k8s.io/cloud-provider#Interface)). SCS implements the current cloud provider interface and defines a new one, which SCS cloud providers have to implement. In there, we unify the naming of the annotations and give a pre-defined set of inputs to the function that, for example, creates a load balancer.
```

Comment:
Wouldn't it be much easier to have a mutating admission hook which generates the corresponding annotations from a "standard" annotation?

THat way, we don't need intermediate objects or controllers or whatnot.

Reply:
The proposed approach only requires implementing one function (CreateLoadBalancer) and creates a clear interface that cloud providers can implement. If you have an easier and better-defined approach, I'm of course open to it!

Without spending a lot of time thinking about the mutation hook approach, it sounds quite hacky and not transparent to me.